egs/gigaspeech/ASR/local/compute_fbank_gigaspeech.py

```diff
@@ -0,0 +1,248 @@
+                # time domain.
+                if partition not in ["DEV", "TEST"]:
+                    cut_set = (
+                        cut_set
+                        + cut_set.perturb_speed(0.9)
+                        + cut_set.perturb_speed(1.1)
+                    )
+                cut_set.to_file(raw_cuts_path)
+
+            cuts_path = output_dir / f"cuts_{partition}{ctx_suffix}.jsonl.gz"
+            if cuts_path.is_file():
+                logging.info(
+                    f"{partition} already exists - skipping cutting into "
+                    f"sub-segments."
+                )
+            else:
+                try:
+                    # If we skipped initializing `cut_set` because it exists
+                    # on disk, we'll load it. This helps us avoid re-computing
+                    # the features for different variants of context windows.
+                    cut_set
+                except NameError:
+                    logging.info(f"Reading {partition} raw cuts from disk.")
+                    cut_set = CutSet.from_file(raw_cuts_path)
+                # Note this step makes the recipe different than LibriSpeech:
+                # Since recordings are long, the initial CutSet has very long
+                # cuts with a plenty of supervisions. We cut these into smaller
+                # chunks centered around each supervision, possibly adding
+                # acoustic context.
+                logging.info(
+                    f"About to split {partition} raw cuts into smaller chunks."
+                )
+                cut_set = cut_set.trim_to_supervisions(
+                    keep_overlapping=False,
+                    min_duration=None
+                    if args.context_window <= 0.0
+                    else args.context_window,
+                    context_direction=args.context_direction,
+                )
+                if partition in ["L", "XL"]:
+                    # Before storing manifests in, we want to pre-shuffle them,
+                    # as the sampler won't be able to do it later in an
+                    # efficient manner.
+                    cut_set = cut_set.shuffle()
+
+                if args.precomputed_features:
+                    # Extract the features after cutting large recordings into
+                    # smaller cuts.
+                    # Note:
+                    # we support very efficient "chunked" feature reads with
+                    # the argument `storage_type=ChunkedLilcomHdf5Writer`,
+                    # but we don't support efficient data augmentation and
+                    # feature computation for long recordings yet.
+                    # Therefore, we sacrifice some storage for the ability to
+                    # precompute features on shorter chunks,
+                    # without memory blow-ups.
+                    cut_set = cut_set.compute_and_store_features(
```

Comment:
I thought Piotr was suggesting `compute_and_store_features_batch`
See https://github.com/lhotse-speech/lhotse/issues/452#issuecomment-961873479

Reply:
Yes -- you'd do best to leverage the changes made by @glynpu here: https://github.com/k2-fsa/icefall/pull/100

Reply:
Does this method also work for XL set? And we no longer need on-the-fly feature extraction, nor pre-shuffling by default, right?

Reply:
I am testing the pre-computed-features interface with "L" and "XL" subset. As suggested by protr here https://github.com/lhotse-speech/lhotse/issues/452

Reply:
This method should work for XL -- Lyong ran into some issues before but I suspect it was related to data prep, we should see soon.

You won't need on-the-fly extraction, but it still allows you to save a lot of disk space. You don't need pre-shuffling anymore though -- Lhotse samplers implement a streaming shuffle variant for lazily-opened CutSets (it works with zero code changes when `shuffle=True` is passed to a sampler).

Reply:
Here we do `compute_and_store_features_batch` if GPU card is detected, otherwise we go back to `compute_and_store_features`  or suggest `--precomputed-features False`.

Reply:
Also, I also got this errorwhile installing `kaldifeat` https://github.com/csukuangfj/kaldifeat/issues/12#issue-1030657358

Reply:
> I also got this errorwhile installing kaldifeat csukuangfj/kaldifeat#12 (comment)

So did you succeed in installing kaldifeat?

Reply:
> > I also got this errorwhile installing kaldifeat [csukuangfj/kaldifeat#12](https://github.com/csukuangfj/kaldifeat/issues/12) (comment)
> 
> So did you succeed in installing kaldifeat?

not yet. This error was resolved, but another error occurred:
```
...
    [ 50%] Linking CXX shared library ../../lib/libkaldifeat_core.so
    /usr/bin/ld: cannot find /export/c07/gwang50/miniconda3/envs/k2/lib: File format not recognized
    collect2: error: ld returned 1 exit status
    kaldifeat/csrc/CMakeFiles/kaldifeat_core.dir/build.make:223: recipe for target 'lib/libkaldifeat_core.so' failed
    make[3]: *** [lib/libkaldifeat_core.so] Error 1
    CMakeFiles/Makefile2:340: recipe for target 'kaldifeat/csrc/CMakeFiles/kaldifeat_core.dir/all' failed
    make[2]: *** [kaldifeat/csrc/CMakeFiles/kaldifeat_core.dir/all] Error 2
    CMakeFiles/Makefile2:427: recipe for target 'kaldifeat/python/csrc/CMakeFiles/_kaldifeat.dir/rule' failed
    make[1]: *** [kaldifeat/python/csrc/CMakeFiles/_kaldifeat.dir/rule] Error 2
    Makefile:225: recipe for target '_kaldifeat' failed
    make: *** [_kaldifeat] Error 2
...
```

Reply:
So it is the same error posted in
https://github.com/csukuangfj/kaldifeat/issues/15

Please see the discussions in the above issue.

If you cannot resolve it, I would recommend you to install it using the approach posted by Piotr in
https://github.com/csukuangfj/kaldifeat/issues/12#issue-1030657358

Reply:
I tried 
```
pip install kaldifeat
```
```
python setup.py install
```
and
```
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DCUDNN_LIBRARY_PATH="/export/c07/gwang50/miniconda3/envs/k2/lib" -DCUDNN_INCLUDE_PATH="/export/c07/gwang50/miniconda3/envs/k2/include"
make  _kaldifeat
```
still got the same error.

Reply:
@pzelasko How did you install kaldifeat on CLSP grid? Actually I got 3 errors.
First is the version of `cmake` on grid is 3.7.2, and I installed a version of 3.22.0.
Second is `cannot find CUDNN`, resolved by specifying `DCUDNN_LIBRARY_PATH` and `DCUDNN_INCLUDE_PATH`.
Did you ever get the third error above`/usr/bin/ld: cannot find /export/c07/gwang50/miniconda3/envs/k2/lib: File format not recognized`?

Reply:
@wgb14 
Could you replace
```
-DCUDNN_LIBRARY_PATH="/export/c07/gwang50/miniconda3/envs/k2/lib"
```

with

```
-DCUDNN_LIBRARY_PATH="/export/c07/gwang50/miniconda3/envs/k2/lib/libcudnn.so"
```

I assume that `libcudnn.so` exists in `/export/c07/gwang50/miniconda3/envs/k2/lib`.
If not, please find the path to `libcudnn.so` and replace the above value.

Passing only the directory is not right, I think. We should pass the path to the `.so` file, not to its enclosing directory.

Reply:
cool, it really works. Thank you, Fangjun!

Reply:
Cool! You're probably the first who installed it on the CLSP grid. If you could write down the steps in a gist, we could share it with our students...

Reply:
changed to chunked feature extraction  with GPU by default here, but got OOM error while running on grid
![image](https://user-images.githubusercontent.com/49512813/142094422-4d04f537-2d9c-4e0e-ae20-0be337958e8b.png)
Any insights?

Reply:
Is it possible that the default value of --batch-duration, which is 600, is too large?

Reply:
That is not about GPU memory, it's an error during fork, which is likely due to a limit of virtual memory.  This is probably about the number of dataloader processes (if that's an option), and about the overall size of the manifest.

Reply:
Yeah it's your CPU memory that's blown up. You can try to decrease num workers (there is an option), but I think Dan is right about the manifest size. There is a way to make it load the manifest lazily, but I suggest you start with splitting the cut manifest into N splits and read->process->save each of them sequentially, and then recombine (you can even do that with bash commands gunzip, split, cat, gzip). I will add some code to support CUDA extraction with streaming manifest reads and writes later.

Reply:
... actually, if you're using speed perturbation or other augmentations, ditching them might solve your issues. You can still use MUSAN, SpecAugment, etc. in Dataset later.